### PR TITLE
Add setting to disable aggs optimization

### LINKED ISSUE #73426: Setting to disable rewrite-to-filters optimization
It turns out that *sometimes* my [beautiful date_histogram optimizations](https://www.elastic.co/blog/how-we-made-date-histogram-aggregations-faster-than-ever-in-elasticsearch-7-11) aren't always optimizations. Sometimes they make the `date_histogram` super ultra slow. We've known that and have heuristics that disable the optimization. Those heuristics generally work because, like, we worked on them. But it turns out that they don't work for everyone. Sometimes folks have queries or data or *something* that doesn't work well with the heuristics we have. I'll be filing a separate issue with the cases we've heard about and fix the heuristics. But I think this is a forever problem. We just won't cover all cases in our tests. So we're going to need an escape hatch to disable the optimization entirely. I believe a cluster level setting is the most sensible thing here. Its a "big hammer", but its one that you can easily modify if you hit the issue without restarting anything and you can disable once you've upgraded to a version that includes the fix.

### PR DESCRIPTION
**Pull request by nik9000**
Sometimes our fancy "run this agg as a Query" optimizations end up
slower than running the aggregation in the old way. We know that and use
heuristics to dissable the optimization in that case. But it turns out
that the process of running the heuristics itself can be slow, depending
on the query. Worse, changing the heuristics requires an upgrade, which
means waiting. If the heurisics make a terrible choice folks need a
quick way out. This adds such a way: a cluster level setting that
contains a list of queries that are considered "too expensive" to try
and optimize. If the top level query contains any of those queries we'll
disable the "run as Query" optimization.

The default for this settings is wildcard and term-in-set queries, which
is fairly conservative. There are certainly wildcard and term-in-set
queries that the optimization works well with, but there are other queries
of that type that it works very badly with. So we're being careful.

Better, you can modify this setting in a running cluster to disable the
optimization if we find a new type of query that doesn't work well.

Closes #73426

### CODE REVIEW

**Comment by nik9000** [comment]
I need to reread this again and add a few more tests, but I'd like to get this out there.

**Comment by nik9000** [comment]
I'd like to add an integration test for this in the `terms` aggregator - that'd let us make sure the settings works.

File: server/src/internalClusterTest/java/org/elasticsearch/search/profile/aggregation/AggregationProfilerIT.java [hunk 1]
```diff
@@ -28,8 +31,11 @@
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
+import java.util.concurrent.TimeUnit;
 import java.util.stream.Collectors;
 
+import static io.github.nik9000.mapmatcher.MapMatcher.assertMap;
+import static io.github.nik9000.mapmatcher.MapMatcher.matchesMap;
```
Comment (iverase) [inline server/src/internalClusterTest/java/org/elasticsearch/search/profile/aggregation/AggregationProfilerIT.java:38]: Oops, that should not be here

Comment (nik9000) [reply server/src/internalClusterTest/java/org/elasticsearch/search/profile/aggregation/AggregationProfilerIT.java:38]: No?

**Comment by nik9000** [comment]
It looks like there is another class of issue around these optimization that comes from the query being very very large. We don't copy the query, but we do make its weight many times. I'd expected that these weights would be cached. But that's not always guaranteed. I think in a follow up change we should skip this optimization also based on the size of the query. We *should* also do more to reduce the re-preparation of the weights. That's a little more complicated though. Lucene has some hard won heuristics around what to cache when and we'd have to be careful there. I think it's best to do the simplest thing.

**Comment by nik9000** [comment]
After talking with some folks it turns out that the setting is too precise and difficult to use. And might cause issues on upgrade. And who knows what else. It was clever, but not worth it.

**Pull request closed without merging**